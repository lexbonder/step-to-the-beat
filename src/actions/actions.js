export const saveAccessToken = token => ({
  type: 'SAVE_ACCESS_TOKEN',
  token
})

export const clearAccessToken = () => ({
  type: 'CLEAR_ACCESS_TOKEN'
})

export const logInUser = () => ({
  type: 'LOG_IN_USER',
  loggedIn: true
})

export const logOutUser = () => ({
  type: 'LOG_OUT_USER',
  loggedIn: false
})

export const saveUserName = userName => ({
  type: 'SAVE_USER_NAME',
  userName
})

export const selectSpm = spm => ({
  type: 'SELECT_SPM',
  spm
})

export const selectGenre = genre => ({
  type: 'SELECT_GENRE',
  genre
})

export const saveSpm = spm => ({
  type: 'SAVE_SPM',
  spm
})