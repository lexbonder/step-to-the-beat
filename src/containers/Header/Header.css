header {
  display: grid;
  grid-template-columns: 65px 1fr 65px;
  grid-template-rows: 60px;
  justify-content: center;
  align-items: center;
  border-bottom: 5px solid #404040;
  text-align: center;
  position: fixed;
  top: 0;
  width: 100vw;
  background-color: #262626;
  z-index: 1;
}

.home-button:hover,
.profile-pic:hover {
  cursor: pointer;
}

.home-button,
.page-name {
  color: #4E9AF5;
  font-size: 30px;
}

.profile-pic {
  display: block;
  margin: auto;
  height: 40px;
  border-radius: 100%;
}

.menu-wrapper {
  position: relative;
}

.menu {
  position: absolute;
  right: 0;
  top: 0;
  padding-top: 50px;
  width: 82px;
}

.log-out {
  padding: 10px;
  background-color: #999999;
  color: #404040;
  text-decoration: underline;
}

.log-out:hover,
.log-out:after {
  background-color: #BBBBBB;
  color: #626262;
  cursor: pointer;
}