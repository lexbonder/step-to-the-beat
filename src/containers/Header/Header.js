import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { getUserContent } from '../../firebaseCalls';
import {
  logOutUser,
  clearAccessToken,
  saveAccessToken,
  logInUser,
  saveUser,
  seedsFromFirebase,
  genresFromFirebase, 
  spmsFromFirebase
} from '../../actions/actions';
import PropTypes from 'prop-types';
import './Header.css';

export class Header extends Component {
  constructor(props) {
    super(props);
    this.state = {
      menuOpen: 'hide',
      errorState: ''
    };
  }

  componentDidMount = async () => {
    const { 
      loggedIn, 
      saveUser, 
      logInUser, 
      saveAccessToken 
    } = this.props;
    if (!loggedIn && localStorage.currentUser) {
      const retrieved = localStorage.getItem('currentUser');
      const currentUser = JSON.parse(retrieved);
      
      logInUser();
      saveUser(currentUser.user);
      saveAccessToken(currentUser.accessToken);
      try {
        const userContent = await getUserContent(currentUser.user.id);
        this.reloadUserContent(userContent);
      } catch (error) {
        this.setState({errorState: error.message});
      }
    }
  }

  reloadUserContent = userContent => {
    const { 
      history,
      seedsFromFirebase, 
      genresFromFirebase, 
      spmsFromFirebase 
    } = this.props;
    if (userContent.val()) {
      const { savedGenres, savedSeeds, savedSpms} = userContent.val();
      seedsFromFirebase(Object.values(savedSeeds || {}));
      genresFromFirebase(savedGenres);
      spmsFromFirebase(savedSpms);
    } 
    history.push('/saved-playlists');
  }

  hideLogOut = () => {
    this.setState({menuOpen: 'hide'});
  }

  showLogOut = () => {
    this.setState({menuOpen: 'show'});
  }

  handleLogOut = () => {
    const {logOutUser, clearAccessToken, history} = this.props;
    logOutUser();
    clearAccessToken();
    localStorage.removeItem('currentUser');
    history.push('/');
  }

  handleHomeButton = () => {
    const { history } = this.props;
    history.push('/saved-playlists');
  }
  
  pageName = () => {
    const { pathname } = this.props.location;
    switch (pathname) {
    case '/saved-playlists':
      return 'My Playlists';
    case '/select-spm':
      return 'Steps Per Minute';
    case '/calculations':
      return 'Calculate SPM';
    case '/select-genre':
      return 'Select Genre';
    case '/confirm':
      return 'Confirm';
    case '/playlist':
      return 'View Playlist';
    default:
      return '';
    }
  }

  userGreeting = () => {
    const { user, loggedIn } = this.props;
    if (loggedIn) {
      return (
        <header>
          <h2 
            onClick={this.handleHomeButton}
            className='home-button'
          >
            <i className="fas fa-home"></i>
          </h2>
          <h2 className='page-name'>{this.pageName()}</h2>
          <div 
            className='menu-wrapper'>
            <img
              className='profile-pic'
              src={user.image}
              alt='profile'
            />
            <div 
              className='menu'
              onMouseEnter={this.showLogOut}
              onMouseLeave={this.hideLogOut}>
              <p 
                className={`log-out ${this.state.menuOpen}`}
                onClick={this.handleLogOut}>Log Out</p>
            </div>
          </div>
        </header>
      );
    }
  }

  render() {
    return (
      <div>
        {this.userGreeting()}
      </div>
    );
  }
}

const { bool, shape, string, func } = PropTypes;

Header.propTypes = {
  loggedIn: bool,
  user: shape({
    name: string,
    id: string
  }),
  logOutUser: func,
  clearAccessToken: func,
  history: shape({
    push: func
  }),
  location: shape({
    pathname: string
  }),
  seedsFromFirebase: func,
  genresFromFirebase: func,
  spmsFromFirebase: func
};

export const MSTP = ({ loggedIn, user }) => ({
  loggedIn,
  user
});

export const MDTP = (dispatch) => ({
  logOutUser: () => dispatch(logOutUser()),
  clearAccessToken: () => dispatch(clearAccessToken()),
  logInUser: () => dispatch(logInUser()),
  saveAccessToken: accessToken => dispatch(saveAccessToken(accessToken)),
  saveUser: user => dispatch(saveUser(user)),
  seedsFromFirebase: seeds => dispatch(seedsFromFirebase(seeds)),
  genresFromFirebase: genres => dispatch(genresFromFirebase(genres)),
  spmsFromFirebase: spms => dispatch(spmsFromFirebase(spms))
});

export default withRouter(connect(MSTP, MDTP)(Header));