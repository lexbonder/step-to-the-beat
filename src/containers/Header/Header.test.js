/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { Header, MSTP, MDTP } from './Header';
import { logOutUser, clearAccessToken } from '../../actions/actions';

describe('Header', () => {
  let wrapper;
  let mockLoggedIn;
  let mockUser = {name: 'Alex', id: 'lxbndr'}
  let mockLogOutUser = jest.fn()
  let mockClearAccessToken = jest.fn()
  let mockHistory = { push: jest.fn() }
  let mockLocation = { pathname: '/confirm'}
  window.localStorage = {
    getItem: jest.fn(),
    setItem: jest.fn(),
    removeItem: jest.fn()
  }

  beforeEach(() => {
    mockLoggedIn = true
    wrapper = shallow(<Header 
      loggedIn={mockLoggedIn}
      user={mockUser}
      logOutUser={mockLogOutUser}
      clearAccessToken={mockClearAccessToken}
      history={mockHistory}
      location={mockLocation}
    />)
  })

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual({menuOpen: 'hide'})
  })

  describe('While a user is logged in', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })
    
    describe('userGreeting', () => {
      it('should match the snapshot if a user is logged in', () => {
        expect(wrapper.instance().userGreeting()).toMatchSnapshot()
      })

      it('should call logOutUser and clearAccessToken when a user clicks the LogOut link', () => {
        wrapper.find('.log-out').simulate('click')

        expect(mockLogOutUser).toHaveBeenCalled()
        expect(mockClearAccessToken).toHaveBeenCalled()
      })
    })
  
  })

  describe('While a user is not logged in', () => {
    beforeEach(() => {
      mockLoggedIn = false
      wrapper = shallow(<Header 
        loggedIn={mockLoggedIn}
        user={mockUser}
        logOutUser={mockLogOutUser}
        clearAccessToken={mockClearAccessToken}
        history={mockHistory}
        location={mockLocation}
      />)
    })

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    describe('userGreeting', () => {
      it('should return undefined', () => {
        expect(wrapper.instance().userGreeting()).toEqual(undefined)
      })
    })

  })

  describe('showLogOut', () => {
    it('should set menuOpen to show', () => {
      wrapper.instance().showLogOut()
      expect(wrapper.state().menuOpen).toEqual('show')
    })
  })

  describe('hideLogOut', () => {
    it('should set menuOpen to hide if it is set to show', () => {
      wrapper.instance().hideLogOut()
      expect(wrapper.state().menuOpen).toEqual('hide')
    })
  })

  describe('pageName', () => {
    it('should return Confirm when the pathname is /confirm', () => {
      expect(wrapper.instance().pageName()).toEqual('Confirm')
    })

    it('should return My Playilsts when the pathname is /saved-playlists', () => {
      const mockLocation = {pathname: '/saved-playlists'}
      const wrapper = shallow(<Header 
        loggedIn={mockLoggedIn}
        user={mockUser}
        logOutUser={mockLogOutUser}
        clearAccessToken={mockClearAccessToken}
        history={mockHistory}
        location={mockLocation}
      />)
      expect(wrapper.instance().pageName()).toEqual('My Playlists')
    })

    it('should return Steps Per Minute when the pathname is /select-spm', () => {
      const mockLocation = {pathname: '/select-spm'}
      const wrapper = shallow(<Header 
        loggedIn={mockLoggedIn}
        user={mockUser}
        logOutUser={mockLogOutUser}
        clearAccessToken={mockClearAccessToken}
        history={mockHistory}
        location={mockLocation}
      />)
      expect(wrapper.instance().pageName()).toEqual('Steps Per Minute')
    })

    it('should return Select Genre when the pathname is /select-genre', () => {
      const mockLocation = {pathname: '/select-genre'}
      const wrapper = shallow(<Header 
        loggedIn={mockLoggedIn}
        user={mockUser}
        logOutUser={mockLogOutUser}
        clearAccessToken={mockClearAccessToken}
        history={mockHistory}
        location={mockLocation}
      />)
      expect(wrapper.instance().pageName()).toEqual('Select Genre')
    })

    it('should return View Playlist when the pathname is /playlist', () => {
      const mockLocation = {pathname: '/playlist'}
      const wrapper = shallow(<Header 
        loggedIn={mockLoggedIn}
        user={mockUser}
        logOutUser={mockLogOutUser}
        clearAccessToken={mockClearAccessToken}
        history={mockHistory}
        location={mockLocation}
      />)
      expect(wrapper.instance().pageName()).toEqual('View Playlist')
    })

    it('should return an empty string by default', () => {
      const mockLocation = {pathname: '/nonsense'}
      const wrapper = shallow(<Header 
        loggedIn={mockLoggedIn}
        user={mockUser}
        logOutUser={mockLogOutUser}
        clearAccessToken={mockClearAccessToken}
        history={mockHistory}
        location={mockLocation}
      />)
      expect(wrapper.instance().pageName()).toEqual('')
    })
  })

  describe('handleLogOut', () => {
    it('should call logOutUser, clearAccessToken and reirect the user to / when called', () => {
      wrapper.instance().handleLogOut()

      expect(mockLogOutUser).toHaveBeenCalled()
      expect(mockClearAccessToken).toHaveBeenCalled()
      expect(mockHistory.push).toHaveBeenCalledWith('/')
    })
  })

    describe('handleHomeButton', () => {
    it('should redirect the user to /saved-playlists when called', () => {
      wrapper.instance().handleHomeButton()
      expect(mockHistory.push).toHaveBeenCalledWith('/saved-playlists')
    })
  })

  describe('MSTP', () => {
    it('should return an object with the values stored in state', () => {
      const mockState = {
        loggedIn: true,
        user: {name: 'Alex', id: 'lxbndr'}
      }
      const mapped = MSTP(mockState)

      expect(mapped.loggedIn).toEqual(true)
      expect(mapped.user).toEqual({name: 'Alex', id: 'lxbndr'})
    })
  })

  describe('MDTP', () => {
    it('should call dispatch when one of MDTPs funcitons are called',() => {
      const mockDispatch = jest.fn()
      const result = MDTP(mockDispatch)
      result.logOutUser()
      result.clearAccessToken()

      expect(mockDispatch).toHaveBeenCalledWith(logOutUser())
      expect(mockDispatch).toHaveBeenCalledWith(clearAccessToken())
    })
  })

})