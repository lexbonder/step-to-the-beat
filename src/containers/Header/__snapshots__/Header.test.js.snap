// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header While a user is logged in should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header
    clearAccessToken={[Function]}
    history={
        Object {
            "push": [Function],
          }
    }
    location={
        Object {
            "pathname": "/confirm",
          }
    }
    logOutUser={[Function]}
    loggedIn={true}
    user={
        Object {
            "id": "lxbndr",
            "name": "Alex",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <header>
        <button
                className="home-button"
                onClick={[Function]}
        >
                <i
                        className="fas fa-home"
                />
        </button>
        <h2
                className="page-name"
        >
                Confirm
        </h2>
        <div
                className="menu-wrapper"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
        >
                <img
                        alt="profile"
                        className="profile-pic"
                        src={undefined}
                />
                <div
                        className="menu hide"
                >
                        <p
                                className="log-out"
                                onClick={[Function]}
                        >
                                Log Out
                        </p>
                </div>
        </div>
</header>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="home-button"
            onClick={[Function]}
>
            <i
                        className="fas fa-home"
            />
</button>,
          <h2
            className="page-name"
>
            Confirm
</h2>,
          <div
            className="menu-wrapper"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
>
            <img
                        alt="profile"
                        className="profile-pic"
                        src={undefined}
            />
            <div
                        className="menu hide"
            >
                        <p
                                    className="log-out"
                                    onClick={[Function]}
                        >
                                    Log Out
                        </p>
            </div>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <i
              className="fas fa-home"
/>,
            "className": "home-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fas fa-home",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Confirm",
            "className": "page-name",
          },
          "ref": null,
          "rendered": "Confirm",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="profile"
                className="profile-pic"
                src={undefined}
/>,
              <div
                className="menu hide"
>
                <p
                                className="log-out"
                                onClick={[Function]}
                >
                                Log Out
                </p>
</div>,
            ],
            "className": "menu-wrapper",
            "onMouseEnter": [Function],
            "onMouseLeave": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "profile",
                "className": "profile-pic",
                "src": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <p
                  className="log-out"
                  onClick={[Function]}
>
                  Log Out
</p>,
                "className": "menu hide",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Log Out",
                  "className": "log-out",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Log Out",
                "type": "p",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "header",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <header>
          <button
                    className="home-button"
                    onClick={[Function]}
          >
                    <i
                              className="fas fa-home"
                    />
          </button>
          <h2
                    className="page-name"
          >
                    Confirm
          </h2>
          <div
                    className="menu-wrapper"
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
          >
                    <img
                              alt="profile"
                              className="profile-pic"
                              src={undefined}
                    />
                    <div
                              className="menu hide"
                    >
                              <p
                                        className="log-out"
                                        onClick={[Function]}
                              >
                                        Log Out
                              </p>
                    </div>
          </div>
</header>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="home-button"
              onClick={[Function]}
>
              <i
                            className="fas fa-home"
              />
</button>,
            <h2
              className="page-name"
>
              Confirm
</h2>,
            <div
              className="menu-wrapper"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
>
              <img
                            alt="profile"
                            className="profile-pic"
                            src={undefined}
              />
              <div
                            className="menu hide"
              >
                            <p
                                          className="log-out"
                                          onClick={[Function]}
                            >
                                          Log Out
                            </p>
              </div>
</div>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fas fa-home"
/>,
              "className": "home-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fas fa-home",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Confirm",
              "className": "page-name",
            },
            "ref": null,
            "rendered": "Confirm",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <img
                  alt="profile"
                  className="profile-pic"
                  src={undefined}
/>,
                <div
                  className="menu hide"
>
                  <p
                                    className="log-out"
                                    onClick={[Function]}
                  >
                                    Log Out
                  </p>
</div>,
              ],
              "className": "menu-wrapper",
              "onMouseEnter": [Function],
              "onMouseLeave": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "profile",
                  "className": "profile-pic",
                  "src": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <p
                    className="log-out"
                    onClick={[Function]}
>
                    Log Out
</p>,
                  "className": "menu hide",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Log Out",
                    "className": "log-out",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Log Out",
                  "type": "p",
                },
                "type": "div",
              },
            ],
            "type": "div",
          },
        ],
        "type": "header",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Header While a user is logged in userGreeting should match the snapshot if a user is logged in 1`] = `
<header>
  <button
    className="home-button"
    onClick={[Function]}
  >
    <i
      className="fas fa-home"
    />
  </button>
  <h2
    className="page-name"
  >
    Confirm
  </h2>
  <div
    className="menu-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <img
      alt="profile"
      className="profile-pic"
      src={undefined}
    />
    <div
      className="menu hide"
    >
      <p
        className="log-out"
        onClick={[Function]}
      >
        Log Out
      </p>
    </div>
  </div>
</header>
`;

exports[`Header While a user is not logged in should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Header
    clearAccessToken={[Function]}
    history={
        Object {
            "push": [Function],
          }
    }
    location={
        Object {
            "pathname": "/confirm",
          }
    }
    logOutUser={[Function]}
    loggedIn={false}
    user={
        Object {
            "id": "lxbndr",
            "name": "Alex",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
